### PR TITLE
Fix for #538

### PR DESCRIPTION
**Pull request by IkonOne**
This pull addresses the issue outlined in #538.

I couldn't think of any reason that Tilemap should assume that TileAtlas is a gridded tileset with it's origin at (0,0) instead of simply using the regions defined in the atlas.

**Change overview.**

* Moved _tileWidth and _tileHeight from Tilemap to TileAtlas.
* Removed _setCount, _setRows, _setColumns and _tile from Tilemap.
* Changed Tilemap.drawTile() uses data from the AtlasRegion instead of _tile.

**Public API changes:**

* Removed updateTileRect(), getIndex(), getX() and getY() from Tilemap.

### CODE REVIEW

**Comment by bendmorris** [comment]
Seems reasonable to me.

Tilemap's constructor still requires a `tileWidth` and `tileHeight`, but if you're passing in a TileAtlas which has defined its tiles already, they'll be silently ignored. That seems confusing. I guess the current situation isn't any better, as you could pass in a different tileWidth/tileHeight for the same asset and the second one would be ignored now. Seems like something we should fix. Maybe `tileWidth` and `tileHeight` should be optional, and throw an error if they're not provided and the tiles aren't already defined. What do you think?

File: haxepunk/graphics/tile/Tilemap.hx [hunk 1]
```diff
@@ -497,9 +461,10 @@ class Tilemap extends Graphic
 
 	function drawTile(tile:Int, tx:Int, ty:Int, x:Float, y:Float, scx:Float, scy:Float)
 	{
-		updateTileRect(tile);
+		var region = _atlas.getRegion(tile);
+		HXP.rect.setTo(region.x, region.y, region.width, region.height);
 		_atlas.prepareTile(
-			_tile,
+			HXP.rect,
```
Comment (bendmorris) [inline haxepunk/graphics/tile/Tilemap.hx]: I'd rather Tilemap have its own static Rectangle than use HXP.rect for this. If enough classes use HXP.rect it's hard to reason about whether any of their modifications might overlap. This has caused weird bugs in the past.

Comment (IkonOne) [reply haxepunk/graphics/tile/Tilemap.hx]: That makes sense to me.  I'll add the _tile rect back in.

**Comment by IkonOne** [comment]
@bendmorris That looks good to me as well.

Additional changes:

* Move tileMargin/Spacing to the TileAtlas
* in Tilemap constructor.
```
    if TileAtlas.tileWidth/Height <= 0
        if param.tileWidth/Height <= 0
            throw error
        else
            TileAtlas.prepare
```

Thanks for the feedback.  That helped me think through the changes much better.

**Comment by bendmorris** [comment]
Thanks!

**Pull request merged**